vpr/src/base/vpr_constraints_serializer.h

```diff
@@ -0,0 +1,232 @@
+#ifndef VPR_CONSTRAINTS_SERIALIZER_H_
+#define VPR_CONSTRAINTS_SERIALIZER_H_
+
+#include "region.h"
+#include "vpr_constraints.h"
+#include "partition.h"
+#include "partition_region.h"
+#include "echo_files.h"
+#include "constraints_load.h"
+#include "vtr_log.h"
+#include "globals.h" //for the g_vpr_ctx
+
+#include "vpr_constraints_uxsdcxx_interface.h"
+
+struct VprConstraintsContextTypes : public uxsd::DefaultVprConstraintsContextTypes {
+    using AddAtomReadContext = void*;
+    using AddRegionReadContext = void*;
+    using PartitionReadContext = void*;
+    using PartitionListReadContext = void*;
+    using VprConstraintsReadContext = void*;
+    using AddAtomWriteContext = void*;
+    using AddRegionWriteContext = void*;
+    using PartitionWriteContext = void*;
+    using PartitionListWriteContext = void*;
+    using VprConstraintsWriteContext = void*;
+};
+
+class VprConstraintsSerializer final : public uxsd::VprConstraintsBase<VprConstraintsContextTypes> {
+  public:
+    void start_load(const std::function<void(const char*)>* report_error_in) final {
+        // report_error_in should be invoked if VprConstraintsSerializer encounters
+        // an error during the read.
+        report_error_ = report_error_in;
+    }
+
+    virtual void start_write() final {}
+    virtual void finish_write() final {}
+
+    // error_encountered will be invoked by the reader implementation whenever
+    // any error is encountered.
+    //
+    // This method should **not** be invoked from within VprConstraintsSerializer,
+    // instead the error should be reported via report_error.  This enables
+    // the reader implementation to add context (e.g. file and line number).
+    void error_encountered(const char* file, int line, const char* message) final {
+        vpr_throw(VPR_ERROR_OTHER, file, line, "%s", message);
+    }
+
+    /** Generated for complex type "add_atom":
+     * <xs:complexType name="add_atom">
+     *   <xs:attribute name="name_pattern" type="xs:string" use="required" />
+     * </xs:complexType>
+     */
+    virtual inline const char* get_add_atom_name_pattern(void*& /*ctx*/) final {
+        return temp_.c_str();
+    }
+
+    virtual inline void set_add_atom_name_pattern(const char* name_pattern, void*& /*ctx*/) final {
+        auto& atom_ctx = g_vpr_ctx.atom();
+        std::string atom_name = name_pattern;
+        atom_id_ = atom_ctx.nlist.find_block(name_pattern);
```

Comment:
Have a warning if you encounter name_pattern that doesnt match anything

Reply:
comment what happens when invalid

Reply:
for regex -> atom_id_ to a vector of atom ids because you may get multiple atom ids that match a regular expression

Reply:
1. Check for invalid names, check that in. --> Merge after this
2. Check for signal names, check in. --> new branch off master
3. Check for wildcard names, check in.